### DISCUSSION TITLE
Doubt how to use Validator AdonisJs 6

### DISCUSSION THREAD

**Discussion by CRodolfosa** [adonisjs/core#4550] [General]
Hi, I'm very new to AdonisJs,
I'm following an Adonis5 tutorial and implementing it on Adonis 6.

https://adocasts.com/lessons/adonisjs-user-role-authentication-in-15-minutes

https://github.com/adocasts/user-role-authentication-example/blob/main/app/Controllers/Http/AuthController.ts

I want to create user authentication and Rules using access Token in Adonis 6,
However, the tutorial is on an inferior Adonis and has some changes.

I'm not sure how to create the example validators for my application.
example app:

```
import type { HttpContextContract } from '@ioc:Adonis/Core/HttpContext'
import { schema, rules } from '@ioc:Adonis/Core/Validator'
import User from 'App/Models/User'

export default class AuthController {
  public async register({ request, response, auth }: HttpContextContract) {
    const userSchema = schema.create({
      email: schema.string([rules.email(), rules.trim()]),
      password: schema.string([rules.minLength(8)])
    })

    const data = await request.validate({ schema: userSchema })
    const user = await User.create(data)

    await auth.login(user)
    await user.sendVerifyEmail()

    return response.redirect().toPath('/')
  }
```
nessa versão importa 
In this version, the validator is used like this:
```
import { schema, rules } from '@ioc:Adonis/Core/Validator'
```

I'm trying to follow this doc:

https://docs.adonisjs.com/guides/validation

How does my make:validator look?
node ace make:validator schema
node ace make:validator roles

Can anyone guide me?

**Comment by tomgobich** [comment]
Hi CRodolfosa, welcome to AdonisJS!

So the `node ace make:validator` command will create a validator file inside your project. Inside this file, you can then define one or more validators. In AdonisJS 5, the `schema` and `rules` objects imported from `@ioc:Adonis/Core/Validator` were tools to define and build a validator. With `schema` you could define the structure of the validator (what the object looked like), so you could create a validator with `schema.create({ ... })`, specify the email should be a string via `schema.string()`, etc. Then, `rules` allowed you to define specific rules for those properties like email should be a valid email via `rules.email()`.

In AdonisJS 6, `schema` and `rules` still exist, they're just now all within the `vine` object imported from `@vinejs/vine`. `vine.compile` nested with `vine.object` replaces `schema.create` to allow more flexibility, and rules are now included directly within the `vine` object.

So, since the validator shared above is for a authentication, you could create a validator file called auth where we can put all validators applicable to authentication within via: 
```bash
node ace make:validator auth
```

Then, within this validator, we can define a specific validator for registration:
```ts
import vine from '@vinejs/vine'

export const registrationValidator = vine.compile(
  vine.object({
    email: vine.string().email().trim(),
    password: vine.string().minLength(8),
  })
)
```

I also have some updated lessons within AdonisJS 6 and VineJS discussing validation as well here:
https://adocasts.com/series/lets-learn-adonisjs-6/lessons/validating-form-data-with-vinejs

Hope this helps!
